lib/elixir/lib/string_io.ex

```diff
@@ -359,6 +337,25 @@ defmodule StringIO do
 
   ## helpers
 
+  defp capture_prompt(%{capture_prompt: false} = s, _prompt), do: s
+  defp capture_prompt(%{capture_prompt: true, output: output} = s, prompt) do
+    %{s | output: <<output::binary, IO.chardata_to_string(prompt)::binary>>}
+  end
+
+  defp binary_collect("", _, acc), do: {acc, ""}
+  defp binary_collect(<<"\r\n"::binary, tail::binary>>, _, acc), do: {<<acc::binary, "\n">>, tail}
+  defp binary_collect(<<"\n"::binary, tail::binary>>, _, acc), do: {<<acc::binary, "\n">>, tail}
+
+  defp binary_collect(<<head::utf8, tail::binary>>, :unicode, acc) do
+    binary_collect(tail, :unicode, <<acc::binary, head::utf8 >>)
+  end
+
+  defp binary_collect(<<head, tail::binary>>, :latin1, acc) do
+    binary_collect(tail, :latin1, <<acc::binary, head >>)
+  end
+
+  defp binary_collect(<<_::binary>>, _, _), do: :error
```

Comment:
I've made this change in 20b1df39515a99bad31ab190ca6a3f2f7b90a176.

I guess I should have known to do that since you gave me the exact same advice in a [different PR](https://github.com/elixir-lang/elixir/pull/6044#issuecomment-298150724) a few days ago 😄 . This change definitely improves the performance, especially when reading large lines. When I run the benchmark with each line containing 5,000 characters then this change results in a 25% reduction in time.

In the `split_at_eol` function I'm dealing with lines ending in `"\r\n"`. It's a little ugly, but it seems fast. Please let me know if you have any suggestions about how to improve this.